### PR TITLE
Changing to unique pointers for features

### PR DESCRIPTION
**Pull request by ovysotska**
* Remove share pointers for feature folder
* Respective changes to database and relocalizer

### CODE REVIEW

File: src/localization/features/cnn_feature.cpp [hunk 1]
```diff
@@ -69,20 +72,15 @@ void CnnFeature::binarize() {
...
 
-double CnnFeature::computeSimilarityScore(const iFeature::ConstPtr &rhs) const {
-  const auto featurePtr = std::static_pointer_cast<const CnnFeature>(rhs);
-  if (!featurePtr) {
-    printf("[ERROR][Feature] It seems like you are trying to match features of "
-           "different type\n");
-    exit(EXIT_FAILURE);
-  }
+double CnnFeature::computeSimilarityScore(const iFeature &rhs) const {
+  const auto &feature = static_cast<const CnnFeature&>(rhs);
```
Comment (niosus) [inline src/localization/features/cnn_feature.cpp:76]: I would say that de-facto now the code only works if you actually provide the `CnnFeature` as an `iFeature` here. If the type is different the behavior seems to be undefined. It seems like we're missing some way to work with an `iFeature` in a polymorphic manner to compare it to other features. We could do one of the following here:
- Add a method `type()` to the `iFeature` interface that would return the type of each feature object and then only perform the code in _this_ function if the types match
- Add a method `vector()` to the `iFeature` interface to ensure that we can get the values of the feature in a uniform way.

File: src/localization/features/feature_buffer.cpp [hunk 2]
```diff
@@ -46,21 +47,12 @@ void FeatureBuffer::deleteFeature() {
...
 
-void FeatureBuffer::addFeature(int id, const iFeature::ConstPtr &feature) {
-  if (ids.size() > static_cast<size_t>(std::numeric_limits<int>::max())) {
-    fprintf(stderr,
-            "[ERROR] ids vector size does not fit in integer type. Cannot add "
-            "feature with id: %d.\n",
-            id);
-    return;
-  }
+void FeatureBuffer::addFeature(int id,  std::unique_ptr<iFeature> feature) {
```
Comment (niosus) [inline src/localization/features/feature_buffer.cpp:50]: You might want to get an rref here, i.e., 
```cpp
void FeatureBuffer::addFeature(int id,  std::unique_ptr<iFeature>&& feature) {
```
This is what anyway _is_ happening under the hood as the compiler is able to figure out that you want an rref here, so we might as well be explicit about it.

File: src/localization/relocalizers/lsh_cv_hashing.cpp [hunk 3]
```diff
@@ -60,16 +60,16 @@ void LshCvHashing::train(std::vector<iFeature::Ptr> features) {
...
-    feature.at<uchar>(0, i) = fPtr->bits[i];
+  cv::Mat featureCV(1, feature.bits.size(), CV_8UC1);
+  for (int i = 0; i < feature.bits.size(); ++i) {
+    featureCV.at<uchar>(0, i) = feature.bits[i];
   }
   // _matcherPtr->match(feature, matches);
   Timer timer;
   timer.start();
-  _matcherPtr->knnMatch(feature, matches, 5);
+  _matcherPtr->knnMatch(featureCV, matches, 5);
```
Comment (niosus) [inline src/localization/relocalizers/lsh_cv_hashing.cpp:72]: nit: starting variables with an `_` is usually reserved for the STL code. I would probably move the `_` to the end of the variable.

File: src/localization/relocalizers/lsh_cv_hashing.cpp [hunk 4]
```diff
@@ -97,17 +97,17 @@ std::vector<int> LshCvHashing::getCandidates(int quId) {
     exit(EXIT_FAILURE);
   }
   printf("Getting candidates for a query image\n");
-  const auto featurePtr = std::static_pointer_cast<const iFeature>(
+  const auto &feature = static_cast<const iFeature&>(
```
Comment (niosus) [inline src/localization/relocalizers/lsh_cv_hashing.cpp:100]: Do you need this cast here?

File: src/localization/relocalizers/lsh_cv_hashing.cpp [hunk 5]
```diff
@@ -97,17 +97,17 @@ std::vector<int> LshCvHashing::getCandidates(int quId) {
...
       _database->getQueryFeature(quId));
-  if (!featurePtr) {
-    printf("Wrong feature format\n");
-  }
+  // if (!featurePtr) {
```
Comment (niosus) [inline src/localization/relocalizers/lsh_cv_hashing.cpp:102]: nit: it's usually better to remove the comments and rely on the versioning system to maintain the history of the code changes. Did you mean to leave this comment here?

File: src/localization/relocalizers/lsh_cv_hashing.h [hunk 6]
```diff
@@ -27,6 +27,7 @@
 #include "database/online_database.h"
 #include "features/ifeature.h"
 #include "relocalizers/irelocalizer.h"
+#include <memory>
```
Comment (niosus) [inline src/localization/relocalizers/lsh_cv_hashing.h:30]: nit: It seems to be a nice style to separate the local includes from the system ones by 1 empty line.

**Review by niosus — COMMENTED** [review]
Overall a very nice change. I find that the code is much cleaner now that it does not rely on `shared_ptr` and the ownership is very nicely maintainer. Great work!

My only real concern is about the `static_cast` that assumes the feature type as it might hit us in the back in the long run, but other than that great job! 👍

File: src/localization/relocalizers/lsh_cv_hashing.cpp (hunk 3 above)
Comment (ovysotska) [reply src/localization/relocalizers/lsh_cv_hashing.cpp:72]: Yeap, this is going to be changed in the next iterations when I am going to refactor the relocalizers.

File: src/apps/cost_matrix_based_matching/cost_matrix_based_matching_lsh.cpp [hunk 7]
```diff
@@ -39,21 +39,20 @@
...
+std::vector<std::unique_ptr<iFeature>> loadFeatures(const std::string &path2folder) {
   LOG(INFO) << "Loading the features to hash with LSH.";
   std::vector<std::string> featureNames = listProtoDir(path2folder, ".Feature");
-  std::vector<iFeature::Ptr> featurePtrs;
+  std::vector<std::unique_ptr<iFeature>> features;
 
   for (size_t i = 0; i < featureNames.size(); ++i) {
-    iFeature::Ptr featurePtr = iFeature::Ptr(new CnnFeature(featureNames[i]));
-    featurePtrs.push_back(featurePtr);
+    std::unique_ptr<iFeature> feature = std::make_unique<CnnFeature>(featureNames[i]);
```
[the remark this answers is not in the document]
Comment (ovysotska) [reply src/apps/cost_matrix_based_matching/cost_matrix_based_matching_lsh.cpp:49]: great idea, thanks!

File: src/localization/relocalizers/lsh_cv_hashing.h (hunk 6 above)
Comment (ovysotska) [reply src/localization/relocalizers/lsh_cv_hashing.h:30]: done.

File: src/localization/database/online_database.cpp [hunk 8]
```diff
@@ -36,16 +36,15 @@
...
 addFeatureIfNeeded(FeatureBuffer &featureBuffer,
                    const std::vector<std::string> &featureNames,
                    FeatureType type, int featureId) {
   if (featureBuffer.inBuffer(featureId)) {
     return featureBuffer.getFeature(featureId);
   }
-  const auto feature = createFeature(type, featureNames[featureId]);
-  featureBuffer.addFeature(featureId, feature);
-  return feature;
+  featureBuffer.addFeature(featureId, createFeature(type, featureNames[featureId]));
```
[the remark this answers is not in the document]
Comment (ovysotska) [reply src/localization/database/online_database.cpp:46]: Ah. maybe next time :)

File: src/localization/relocalizers/lsh_cv_hashing.cpp (hunk 4 above)
Comment (ovysotska) [reply src/localization/relocalizers/lsh_cv_hashing.cpp:100]: Actually not. Thanks for catching this.

File: src/localization/features/cnn_feature.cpp (hunk 1 above)
Comment (ovysotska) [reply src/localization/features/cnn_feature.cpp]: For now, I added a type as a field in the class and the dimensions to be able to access the values from every type.

File: src/localization/relocalizers/lsh_cv_hashing.cpp (hunk 5 above)
Comment (ovysotska) [reply src/localization/relocalizers/lsh_cv_hashing.cpp]: Neh, not really. Just forgot. Thanks.

**Pull request merged**